src/sinks/util/tcp.rs

```diff
@@ -37,82 +39,182 @@ use tokio_util::codec::{BytesCodec, FramedWrite};
 #[serde(deny_unknown_fields)]
 pub struct TcpSinkConfig {
     pub address: String,
-    pub encoding: EncodingConfig<Encoding>,
     pub tls: Option<TlsConfig>,
 }
 
+#[derive(Clone)]
+struct TcpConnector {
+    host: String,
+    port: u16,
+    resolver: Resolver,
+    tls: MaybeTlsSettings,
+}
+
+#[derive(Debug, Snafu)]
+pub enum TcpError {
+    #[snafu(display("Connect error: {}", source))]
+    ConnectError {
+        source: TlsError,
+    },
+    #[snafu(display("Unable to resolve DNS: {}", source))]
+    DnsError {
+        source: crate::dns::DnsError,
+    },
+    #[snafu(display("No addresses returned."))]
+    NoAddresses,
+    SendError {
+        source: tokio::io::Error,
+    },
+}
+
 impl TcpSinkConfig {
-    pub fn new(address: String, encoding: EncodingConfig<Encoding>) -> Self {
-        Self {
-            address,
-            encoding,
-            tls: None,
-        }
+    pub fn new(address: String) -> Self {
+        Self { address, tls: None }
     }
 
-    pub fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+    fn build_connector(&self, cx: SinkContext) -> crate::Result<TcpConnector> {
         let uri = self.address.parse::<http::Uri>()?;
 
         let host = uri.host().ok_or(SinkBuildError::MissingHost)?.to_string();
         let port = uri.port_u16().ok_or(SinkBuildError::MissingPort)?;
 
         let tls = MaybeTlsSettings::from_config(&self.tls, false)?;
 
-        let tcp = TcpSink::new(host, port, cx.resolver(), tls);
-        let healthcheck = tcp.healthcheck();
+        let connector = TcpConnector::new(host, port, cx.resolver(), tls);
+
+        Ok(connector)
+    }
+
+    pub fn build_service(&self, cx: SinkContext) -> crate::Result<(TcpService, Healthcheck)> {
+        let connector = self.build_connector(cx)?;
+        let healthcheck = connector.healthcheck();
+        Ok((connector.into(), healthcheck))
+    }
+
+    pub fn build(
+        &self,
+        cx: SinkContext,
+        encoding: EncodingConfig<Encoding>,
+    ) -> crate::Result<(VectorSink, Healthcheck)> {
+        let connector = self.build_connector(cx.clone())?;
+        let healthcheck = connector.healthcheck();
+        let sink: TcpSink = connector.into();
+        let sink = StreamSinkOld::new(sink, cx.acker())
+            .with_flat_map(move |event| iter_ok(encode_event(event, &encoding)));
+
+        Ok((VectorSink::Futures01Sink(Box::new(sink)), healthcheck))
+    }
+}
+
+impl TcpConnector {
+    fn new(host: String, port: u16, resolver: Resolver, tls: MaybeTlsSettings) -> Self {
+        Self {
+            host,
+            port,
+            resolver,
+            tls,
+        }
+    }
+
+    fn connect(&self) -> BoxFuture<'static, Result<TcpOrTlsStream, TcpError>> {
```

Comment:
Any reason why not BoxFuture instead async?

Reply:
To solve lifetime issue. With async function, I don't think we currently have a way to say "the future returned by this function does not borrow self". Because we store connecting future in State, this lifetime condition is required.